Lazier load

phmongeau opened the pull request:
Hi, this is another attempt at solving #67 

The currente implementation with qute://back is a bit slow, especially when using the webengine backend. This makes it much faster by avoiding loading the history untill the tab is focused.

It would probably be possible to avoid creating the webview untill the tab is loaded, but I haven't looked into that yet.

I also added tab-unload and tab-load commands, but tab-load might not be necessary as refreshing the page does the same thing.

The-Compiler commented:
I think that's a nice approach, although like you said, ideally the webview wouldn't be created at all.

FWIW it might take me some time to review this, as my exams are coming closer, so I'll be quite busy until September. In the meantime, can you please look at the issues brought up by [pylint](https://travis-ci.org/qutebrowser/qutebrowser/jobs/399850880), [flake8](https://travis-ci.org/qutebrowser/qutebrowser/jobs/399850881) and while [generating the docs](https://travis-ci.org/qutebrowser/qutebrowser/jobs/399850882)?

phmongeau commented:
No probleme, take your time

I fixed the pylint, flake8, and docs error.

I'm starting to look into not creating the webview until the tab is loaded.
There's a lot of things acessing the webview directly so I might take me a bit of time to find all the places I need to change. I'll see what I can do.

phmongeau commented on qutebrowser/browser/webengine/webenginetab.py:
```diff
@@ -502,6 +502,13 @@ class WebEngineHistory(browsertab.AbstractHistory):
+                None)
+            if idx is None:
+                idx = len(self.to_load) - 1
+            return idx
```
It should be done in webkit too, I forgot to copy it.

When the session is saved, it uses `tab.history.current_idx() == idx` to find the active history item. If the tab is not loaded self._history.currentIndex() returns nothing so there would be not active item.

phmongeau commented on qutebrowser/browser/webengine/webenginetab.py:
```diff
@@ -502,6 +502,13 @@ class WebEngineHistory(browsertab.AbstractHistory):
 
     def current_idx(self):
+        if self.to_load:
+            idx = next(
```
do you mean doing something like?:

    try:
        idx = [i for i, item in enumerate(self.to_load) if item.active][0]
    except IndexError:
        idx = len(self.to_load) - 1

I used next() to make it stop iterating as soon as it finds the active item, but that might be a premature optimisation. And it would probably be more effecient to iterate in reversed order since the active item is more likely to be near the end of the list.

toofar replied:
Oh okay, I hadn't seen that pattern before.

toofar commented:
> Is there a way to also not focus the tab on undo? My current implementation will load the tab on undo, which you probably wouldn't want to do if the tab wasn't focused.

I was going to say it probably wasn't worth it since tabs reload already when undeleted. But since you are already proposing adding a tab-unload command I suppose it isn't too terrible to move a little more toward lazy tab loading, not just lazy session loading.

I think without changing the current implementation too much you would have to change the loop in `TabbedBrowser.undo()` to pass `background=config.val.session.lazy_restore` to `self.tabopen()`. But then of course you have to stop the tab loading. Which would mean adding a deserialize method to tabhistory (right?) then back in undo setting newtab.loaded = False, deserializing the history to TabHistoryItems and passing them to load_history_items and then conditionally telling the tab to load based on the lazy setting.

toofar commented:
@phmongeau Are you going to continue with this PR? In addition to the requested changes there have been some changes to the tab API which this wil need to be rebased on (mostly moving some history stuff into a private class).

toofar commented:
I rebased on master. I think I did it right. I didn't look at addressing the existing feedback.

phmongeau commented:
Hey, sorry I don't have a lot of time to work on this these days.

I had started addressing the feedback a while ago, but didn't get it into a stable enough state to push my changes.

Here's what I got so far:
https://gist.github.com/phmongeau/2df774c113a12bde28299f13181dc341

If I remember correctly, it was pretty usable, but some tests are failing.

rakanalh commented:
@phmongeau are you going to have the time to complete this PR? i can take over if you don't.

phmongeau commented:
@rakanalh I don't think I'll have time to finish this. Feel free to take over.

rakanalh commented:
Cool, @The-Compiler i could address your feedback to the PR and have this ready for review again. Would you be willing to review/merge this if it was completed?

The-Compiler commented:
Yup - though it'll probably take me a while as there's quite a PR backlog. Also note that it'll likely need to be revisited for Qt 5.14 because of https://codereview.qt-project.org/c/qt/qtwebengine/+/258919

rakanalh commented:
Seems like this `setLifecycleState` is still in alpha according to: https://github.com/qt/qtwebengine/tree/v5.14.0-alpha1

I'll have to wait until this public API is released.

The-Compiler commented:
@rakanalh Yeah - whatever new code there is should also work without that new Qt feature, but at some point, support for it should be added, so that should be kept in mind.

rakanalh commented:
@phmongeau could you please invite me as a collaborator to your fork? i think creating my own PR for this would be a waste of the existing discussions here.

phmongeau commented:
@rakanalh sure, done

rakanalh commented on qutebrowser/browser/browsertab.py:
```diff
@@ -525,19 +525,73 @@ def at_bottom(self):
+        self.to_load = []
 
     def __len__(self):
         return len(self._history)
```
Initially, `_history` was used but now i fully rely on `to_load` so that i only return instances of `AbstractTabHistoryItem`. I thought this was a better interface. I had to also hook into `on_load_finished` so that i add history items to `to_load` as the user navigates pages.

whitelynx replied:
I think @The-Compiler's comment here was about the implementation of `__len__`, not about the line you changed.

rakanalh commented on qutebrowser/browser/browsertab.py:
```diff
@@ -734,6 +798,8 @@ def __init__(self, *, win_id, mode_manager, private, parent=None):
             self, parent=self)
         self.backend = None
+        self.load_on_focus = False
+        self.loaded = True
```
Moved those to `AbstractHistory`

rakanalh commented on qutebrowser/browser/browsertab.py:
```diff
@@ -902,6 +968,67 @@ def progress(self):
+
+        self.history.load_items(self.history.to_load)
+        self.history.to_load = []
+        self.loaded = True
```
done

rakanalh commented on qutebrowser/browser/browsertab.py:
```diff
@@ -525,19 +525,73 @@ def at_bottom(self):
+
+        try:
+            user_data = qt_item.userData()
+        except AttributeError:
```
You're right. I introduced backend specific history item classes.

rakanalh commented on qutebrowser/browser/browsertab.py:
```diff
@@ -902,6 +968,67 @@ def progress(self):
+            ''.format(title=self.title(), icon_tag=icon_tag),
+            self._widget.url())
+
+    def setFocus(self):
```
Switched to using `showEvent`

rakanalh commented on qutebrowser/browser/browsertab.py:
```diff
@@ -902,6 +968,67 @@ def progress(self):
+        if self.load_on_focus:
+            self.load()
+
+    def load_history_items(self, entries):
```
I abstracted away the details of `loaded` and `load_on_focus` but kept the function to pass that down to the tab's history instance.

rakanalh commented on qutebrowser/mainwindow/tabwidget.py:
```diff
@@ -72,6 +72,17 @@ def __init__(self, win_id, parent=None):
         self._init_config()
         config.instance.changed.connect(self._init_config)
 
+    def setCurrentIndex(self, idx):
```
Removed

rakanalh commented:
I think i am done addressing the feedback. I'll have to look at some failing tests and see what i broke :blush:

toofar commented:
`:tab-unload` doesn't seem to work anymore. Once that is working try navigating somewhere so there is a history entry, unload the tab, go back, try going forward. (And thanks for looking at this.)

rakanalh commented:
@toofar my bad... things should be working well now. Let me know if you're facing any issues.

I will continue to work on fixing failing tests meanwhile.

rakanalh commented on qutebrowser/browser/browsertab.py:
```diff
@@ -618,6 +622,42 @@ def load_items(self, items: typing.Sequence) -> None:
+        user_data: The user data for this item.
+    """
+
+    def __init__(self, url: QUrl, title: str, *, original_url: QUrl = None,
```
My preference would be to split each param into it's own line... but i followed convention here.

arza-zara replied:
I also prefer one item per line. I don't think there's a strict convention – even this file has many styles.

rakanalh commented:
@The-Compiler @toofar i believe i have resolved most of the issues / added tests. Will keep an eye on any failing tests for now but i am optimistic about this one.

Would be nice if you can look into the PR and give it a new round of review when you have the time.

rakanalh commented:
The PR fails on scenario: `test_open_with_url` which passes locally.

rakanalh commented:
Alright so regarding the failing test. I was debugging the following scenario:

https://github.com/qutebrowser/qutebrowser/blob/494a8cbad5f9249dcc90458f68081bf06af4391f/tests/end2end/features/open.feature#L5-L17

By running:
```
/bin/tox -e py37 -- -s tests/end2end/features/test_open_bdd.py::test_open_with_url 
```
To force using webkit.

What happens is tab on which `Given I open about:blank` is executed and triggers session auto-save with looks eventually like:
```
{
  "windows": [
    {
      "active": true,
      "geometry": "x01xd9xd0xcbx00x03x00x00x00x00x002x00x00x002x00x00x03Qx00x00x02x89x00x00x002x00x00x002x00x00x03Qx00x00x02x89x00x00x00x00x00x00x00x00x03 x00x00x002x00x00x002x00x00x03Qx00x00x02x89",
      "tabs": [
        {
          "history": [
            {
              "url": "about:blank",
              "title": "",
              "active": true,
              "zoom": 1.0,
              "scroll-pos": {
                "x": 0,
                "y": 0
              },
              "pinned": false
            }
          ],
          "active": true
        },
        {
          "history": [
            {
              "url": "about:blank",
              "title": "",
              "active": true,
              "zoom": 1.0,
              "scroll-pos": {
                "x": 0,
                "y": 0
              },
              "pinned": false
            }
          ]
        }
      ]
    }
  ]
}
```

Until this point everything is correct. However, when:
```
When I run :open http://localhost:(port)/data/numbers/1.txt
```
is executed... the session ends up being:
```
{
  "windows": [
    {
      "active": true,
      "geometry": "x01xd9xd0xcbx00x03x00x00x00x00x002x00x00x002x00x00x03Qx00x00x02x89x00x00x002x00x00x002x00x00x03Qx00x00x02x89x00x00x00x00x00x00x00x00x03 x00x00x002x00x00x002x00x00x03Qx00x00x02x89",
      "tabs": [
        {
          "history": [
            {
              "url": "about:blank",
              "title": "",
              "active": true,
              "zoom": 1.0,
              "scroll-pos": {
                "x": 0,
                "y": 0
              },
              "pinned": false
            }
          ]
        },
        {
          "history": [
            {
              "url": "http://localhost:58573/data/numbers/1.txt",
              "title": "",
              "active": true,
              "zoom": 1.0,
              "scroll-pos": {
                "x": 0,
                "y": 0
              },
              "pinned": false
            }
          ],
          "active": true
        }
      ]
    }
  ]
}
```

That is, before `:tab-only` is executed. Notice that in the second tab, the history seen from the tab's `self.history._history.items()` has only one item which is the new URL that got loaded rather than `about:blank` + the new URL.

I could, by tracing the code, see that `:open` command calls `openurl` function and that in turn calls `curtab.load_url` which eventually calls `self._widget.load(url)` inside the webkit browsertab. 

I am not sure i see how this PR causes this issue. From what i can see, `self._widget.load` is not triggering the widget's history to be updated to include `about:blank` in the first place.

@The-Compiler it would be great if you can shed some light over what i am missing. Thanks!

toofar commented:
@rakanalh Can you elaborate on why you switched to using showEvent instead of when the tab got focussed in https://github.com/qutebrowser/qutebrowser/pull/4037/commits/624b6eb301c8aacf5d895f54eb8657bb0e5984a0? Oh nevermind, I see the boss man said to. Well both setFocus() and setCurrentIndex() are getting called for me, maybe because they are defined as slots? focusInEvent doesn't get called, probably because there is no focus policy.
And what Qt version are you on. The reason I ask is that I have been having an issue that if have a loaded tab to the left (aka at the next lowest index) of a unloaded tab then when I close it the unloaded tab gets loaded, even though it never gets focussed because I have tabs.select_on_remove set to prev.
I've tested on 5.12, 13 and 14. I don't remember if I was having this issue (or using your additions) on 5.11. A traceback in showEvent() shows it is getting called from removeTab() but I don't think it is in our code. I reverted that commit and the problem went away (I haven't thought about other issues or tried it out much apart from to confirm that that particular issue was gone).

rakanalh commented:
@toofar I am familiar with that issue from before this PR. I will look into it and see if i can fix it in a different PR.

As for the rebase, i am using the rebased version of this PR in my integration branch but did not manage to see your issue. But i will come back to this and check it out.

toofar commented:
Okay, it wasn't the rebase that mess it up. It was switching most stuff to use `to_load` even then the tab was loaded. `_can_go_back/forward` where still using `_history` though. I made the following changes to walk back some of the changes to just use `to_load` and a fix for preserving the active item when unloading and to make `__iter__` always return the same item item type which makes the session saving code happy. Also firms up the behaviour when going back/forward on an unloaded tab which was probably pre-existing.

```patch
commit 9f8f63ca4e3a1cb975bf681ea67bfcb6042a4d2c
Author: Jimmy <jimmy@spalge.com>
Date:   Sun Dec 15 11:36:48 2019 +1300

    Don't add to to_load for new history items.
    
    `to_load` holds history items for unloaded tabs, which don't get
    navigation events so no need to add to it here. There was also no
    handling of going back in history and then opening a new page. With the
    backend history implementations gets rid of the future history in that
    case.

diff --git a/qutebrowser/mainwindow/tabbedbrowser.py b/qutebrowser/mainwindow/tabbedbrowser.py
index d3560f514f4f..dd757920c5aa 100644
--- a/qutebrowser/mainwindow/tabbedbrowser.py
+++ b/qutebrowser/mainwindow/tabbedbrowser.py
@@ -360,8 +360,6 @@ class TabbedBrowser(QWidget):
         if not self.is_private:
             web_history = objreg.get('web-history')
             tab.history_item_triggered.connect(web_history.add_from_tab)
-            tab.history_item_triggered.connect(
-                functools.partial(self._on_new_history_item, tab))
 
     def current_url(self):
         """Get the URL of the current tab.
@@ -941,12 +939,6 @@ class TabbedBrowser(QWidget):
             if self.widget.count() == 0:
                 self.tabopen(QUrl('about:blank'))
 
-    def _on_new_history_item(self, tab, url, requested_url, title):
-        history_item = tab.make_tab_history_item(
-            url, title, requested_url, True, None)
-
-        tab.history.to_load.append(history_item)
-
     def resizeEvent(self, e):
         """Extend resizeEvent of QWidget to emit a resized signal afterwards.
 

commit 9a1bd806c3d089500329248b400c4e6b8ef6201c
Author: Jimmy <jimmy@spalge.com>
Date:   Sun Dec 15 11:58:12 2019 +1300

    Preserve active tab flag when unloading.
    
    Otherwise the serializer complains about there being no active entry.
    The subclasses already have this kwarg defined, not sure why it wasn't
    in the (somewhat-)abstract base class.

diff --git a/qutebrowser/browser/browsertab.py b/qutebrowser/browser/browsertab.py
index c95bc7b3b3cb..a2e3b6135bf1 100644
--- a/qutebrowser/browser/browsertab.py
+++ b/qutebrowser/browser/browsertab.py
@@ -724,8 +724,11 @@ class AbstractHistory:
         self.load_on_focus = True
 
         self.to_load = []
+        current_idx = self.current_idx()
         for idx, item in enumerate(self._history.items()):
-            item = self._tab.tab_history_item_from_qt(item)
+            item = self._tab.tab_history_item_from_qt(
+                item, active=idx==current_idx,
+            )
             self.to_load.append(item)
         self._history.clear()
 
@@ -1165,7 +1168,7 @@ class AbstractTab(QWidget):
         if self.history.load_on_focus:
             self.load()
 
-    def tab_history_item_from_qt(self, item):
+    def tab_history_item_from_qt(self, item, active=False):
         raise NotImplementedError
 
     def make_tab_history_item(

commit e4f77f4a0820573996d09c8c6bfabe249de9fedb
Author: Jimmy <jimmy@spalge.com>
Date:   Sun Dec 15 12:00:50 2019 +1300

    Make AbstractHistory defer to backend if not unloaded.
    
    `to_load` wasn't being maintained sufficiently to provide a full history
    backend. Go back to using it as a just a record of what was there when
    the tab was loaded and don't touch use it when the tab is loaded.
    Reverts some of f1af7857ac

diff --git a/qutebrowser/browser/browsertab.py b/qutebrowser/browser/browsertab.py
index a2e3b6135bf1..e8a8f2c17a75 100644
--- a/qutebrowser/browser/browsertab.py
+++ b/qutebrowser/browser/browsertab.py
@@ -661,10 +661,21 @@ class AbstractHistory:
         self.loaded = False
 
     def __len__(self):
-        return len(self.to_load)
+        if self.to_load:
+            return len(self.to_load)
+        else:
+            return len(self._history)
 
     def __iter__(self):
-        return iter(self.to_load)
+        if self.to_load:
+            return iter(self.to_load)
+        else:
+            current_idx = self.current_idx()
+            return (
+                self._tab.tab_history_item_from_qt(item, idx==current_idx)
+                for idx, item in
+                enumerate(self._history.items())
+            )
 
     def _check_count(self, count: int) -> None:
         """Check whether the count is positive."""
@@ -672,10 +683,13 @@ class AbstractHistory:
             raise WebTabError("count needs to be positive!")
 
     def current_idx(self):
-        for i, item in enumerate(self.to_load):
-            if item.active:
-                return i
-        return len(self.to_load) - 1
+        if self.to_load:
+            for i, item in enumerate(self.to_load):
+                if item.active:
+                    return i
+            return len(self.to_load) - 1
+        else:
+            return self._history.currentItemIndex()
 
     def back(self, count: int = 1) -> None:
         """Go back in the tab's history."""

commit 55546e5a8070bb04ec9f03f1684e40681203b5fb
Author: Jimmy <jimmy@spalge.com>
Date:   Sun Dec 15 12:22:23 2019 +1300

    Fix _can_go_forward for unloaded tabs
    
    This is called from the mouse event back forward checking.
    Also brings the webkit implementation in line with the webengine one (at
    this point the non-private backend specific history implementations are
    the same).

diff --git a/qutebrowser/browser/webengine/webenginetab.py b/qutebrowser/browser/webengine/webenginetab.py
index 6f02a5eac782..35423c10c813 100644
--- a/qutebrowser/browser/webengine/webenginetab.py
+++ b/qutebrowser/browser/webengine/webenginetab.py
@@ -610,7 +610,7 @@ class WebEngineHistory(browsertab.AbstractHistory):
         return self.current_idx() > 0
 
     def _can_go_forward(self):
-        return self._history.canGoForward()
+        return self.current_idx() < len(self) - 1
 
     def _item_at(self, i):
         return self._history.itemAt(i)
diff --git a/qutebrowser/browser/webkit/webkittab.py b/qutebrowser/browser/webkit/webkittab.py
index a27cc59b2aa8..4e7087d08a51 100644
--- a/qutebrowser/browser/webkit/webkittab.py
+++ b/qutebrowser/browser/webkit/webkittab.py
@@ -575,10 +575,10 @@ class WebKitHistory(browsertab.AbstractHistory):
         self.private_api = WebKitHistoryPrivate(tab)
 
     def _can_go_back(self):
-        return self._history.canGoBack()
+        return self.current_idx() > 0
 
     def _can_go_forward(self):
-        return self._history.canGoForward()
+        return self.current_idx() < len(self) - 1
 
     def _item_at(self, i):
         return self._history.itemAt(i)

commit a673e6fd9f30e3007434610ab330a7ec1d415757 (HEAD -> pr/4037)
Author: Jimmy <jimmy@spalge.com>
Date:   Sun Dec 15 12:26:11 2019 +1300

    Load unloaded tabs before going back/forwards.
    
    Going backwards was loading something, but losing history. Going forward
    did nothing. I was worried that this would do weird stuff with async
    events and whatnot but it seems good.

diff --git a/qutebrowser/browser/browsertab.py b/qutebrowser/browser/browsertab.py
index e8a8f2c17a75..324ddd42b9e2 100644
--- a/qutebrowser/browser/browsertab.py
+++ b/qutebrowser/browser/browsertab.py
@@ -695,6 +695,8 @@ class AbstractHistory:
         """Go back in the tab's history."""
         self._check_count(count)
         idx = self.current_idx() - count
+        if not self.loaded:
+            self.load()
         if idx >= 0:
             self._go_to_item(self._item_at(idx))
         else:
@@ -705,6 +707,8 @@ class AbstractHistory:
         """Go forward in the tab's history."""
         self._check_count(count)
         idx = self.current_idx() + count
+        if not self.loaded:
+            self.load()
         if idx < len(self):
             self._go_to_item(self._item_at(idx))
         else:

```

rakanalh commented:
Ah that tells me that my rebase was definitely faulty. I will apply your patch and push the change during this week. Thanks @toofar!

rakanalh commented:
Rebased... since pyqtwebengine 5.14 was released, i am going to add support to `setLifecycleState`

blueyed commented:
This is huge!  (as you might know already, just wanted to give some data points / motivation to finish this)

With 247 tabs, on master qutebrowser takes 45s to display the window, and ~3 minutes to report
a blocked request.  (The (current) tab ends up being blank then additionally (maybe only sometimes))

       VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
    9796.9m   2.7g 206680 S   6.6  17.2   2:44.47 qutebrowser

On this branch it display the first tab after 5s already, and uses much less memory:

       VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
    2749876 389692 186612 T   0.0   2.4   0:08.22 qutebrowser

rakanalh commented:
Nice! An update for my side is that i played around with the `setLifecycleState` API and figured out what needs to be done. I still need to sit down and implement this properly. Hoping i can do this during next month.

coiby commented:
> Nice! An update for my side is that i played around with the `setLifecycleState` API and figured out what needs to be done. I still need to sit down and implement this properly. Hoping i can do this during next month.

Hi @rakanalh 

How is your work going? I plan to use the `setLifecycleState` API to implement the feature of discarding a tab after certain time as like  [The Great Suspender](https://chrome.google.com/webstore/detail/the-great-suspender/klbibkeccnjlkjkiokjodocebajanakg). I want to check with you first to avoid duplate effort. 

This is also a feature requested by some users in issue #67. As shown by @blueyed and accroding to [the Qt docs](https://doc.qt.io/qt-5/qtwebengine-features.html#page-lifecycle-api),
> The discarded state is an extreme resource-saving state. In this state, the browsing context of the web view will be discarded and the corresponding renderer subprocess shut down. CPU and memory usage in this state is reduced virtually to zero.

coiby commented:
Hi @phmongeau and @rakanalh, 

While testing out phmongeau's idea that "to avoid creating the webview untill the tab is loaded" in #5370. I'm surprised to find out that the history have been already loaded for each tab after qutebrowser is fully started. Because `setCurrentWidget(tab)` is called for each tab and this triggers showEvent to load the history for all tabs while qutebrowser is still starting. Another surprise is despite that the history are loaded before that any of inactive tabs chosen by the user, no QtWebEngineProcess (type=renderer) is spawned for inactive tabs. Thus it greatly reduce the memory consumption.

I'm curious to figure out how we can load history without spawning QtWebEngineProcess which may simplify the implementation of lazy tab loading. But I can't give a minimum example to load history without spawning QtWebEngineProcess. I wonder if I miss something.

My experiment shows, as long as history is loaded, QtWebEngineProcess would be re-created immediately. Here's the full code

1. First save history to `file.data`
```python
import sys
from PyQt5.Qt import *
from PyQt5.QtWebEngineWidgets import *
from PyQt5.QtWidgets import QApplication

app = QApplication(sys.argv)
web = QWebEngineView()
web.load(QUrl("http://qutebrowser.org/"))
web.show()

app.exec_()
file_ = QFile("file.dat")
file_.open(QIODevice.WriteOnly)
out = QDataStream(file_)
out << web.page().history()
print(web.page().history().count())
file_.close()
```
2. The following code will created 4 tabs loading history from `file.dat`
```python
import sys
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *
from PyQt5.QtWebEngineWidgets import *

class WebView(QWebEngineView):
    def showEvent(self, event: QEvent) -> None:
        file_ = QFile("file.dat")
        file_.open(QIODevice.ReadOnly)
        out = QDataStream(file_)
        his = self.history()
        out >> his
        print(his.count())
        file_.close()

app = QApplication(sys.argv)
tabs = QTabWidget()
url = QUrl("http://example.com")
tabs.show()
for i in range(4):
    web = WebView()
    tabs.addTab(web, str(i))
    tabs.setCurrentWidget(web)
app.exec_()
```

Here's the result of my experiment,
1. In showEvent, after calling `his = self.page().history()`,  QtWebEngineProcess (type=renderer) will be launched
2. after `out >> his`, the created QtWebEngineProcess will be destroyed and re-created again

The result is in accordance with qtwebengine-5.14.2 source code

```c++
// qtwebengine-5.14.2/src/webenginewidgets/api/qwebenginehistory.cpp
QDataStream& operator>>(QDataStream& stream, QWebEngineHistory& history)
{
    history.d_func()->page->recreateFromSerializedHistory(stream);
    return stream;
}

// src/webenginewidgets/api/qwebenginepage.cpp
void QWebEnginePagePrivate::recreateFromSerializedHistory(QDataStream &input)
{
    QSharedPointer<WebContentsAdapter> newWebContents = WebContentsAdapter::createFromSerializedNavigationHistory(input, this);
    if (newWebContents) {
        adapter = std::move(newWebContents);
        adapter->setClient(this);
        adapter->loadDefault();
    }
}

// qtwebengine-5.14.2/src/core/web_contents_adapter.cpp
QSharedPointer<WebContentsAdapter> WebContentsAdapter::createFromSerializedNavigationHistory(QDataStream &input, WebContentsAdapterClient *adapterClient)
{
    int currentIndex;
    std::vector<std::unique_ptr<content::NavigationEntry>> entries;
    deserializeNavigationHistory(input, &currentIndex, &entries, adapterClient->profileAdapter()->profile());

    if (currentIndex == -1)
        return QSharedPointer<WebContentsAdapter>();

    // Unlike WebCore, Chromium only supports Restoring to a new WebContents instance.
    std::unique_ptr<content::WebContents> newWebContents = createBlankWebContents(adapterClient, adapterClient->profileAdapter()->profile());
    content::NavigationController &controller = newWebContents->GetController();
    controller.Restore(currentIndex, content::RestoreType::LAST_SESSION_EXITED_CLEANLY, &entries);
    ...
}
```

Thank you!

rakanalh commented on qutebrowser/browser/webengine/webenginetab.py:
```diff
@@ -1547,6 +1602,32 @@ def _on_load_finished(self, ok: bool) -> None:
                 emit_before_load_started=False))
             self._reload_url = None
 
+    @pyqtSlot(QWebEnginePage.LifecycleState)
```
@The-Compiler I got stuck trying to make this work.

What i am seeing is:
```
TypeError: connect() failed between recommendedStateChanged(QWebEnginePage::LifecycleState) and _check_recommended_lifecycle_state()
```

Which is weird given that the [signal](https://doc.qt.io/qt-5/qwebenginepage.html#recommendedState-prop) is there and matches the signature i am providing.

Any ideas?

toofar replied:
If you look at `_on_feature_permission_requested` for example it points to the enum with a string like `'QWebEnginePage::Feature'`, also make sure you have a new enough pyqt.

rakanalh replied:
I tried using `'QWebEnginePage::LifecycleState'` to no avail. 

>  also make sure you have a new enough pyqt.

Not sure i understand what "enough pyqt" means in this case.

The-Compiler replied:
Might need to do `@pyqtSlot(object)  # QWebEnginePage.LifecycleState` here to keep compatibility with older PyQt versions which don't know about that type.

rakanalh replied:
I did that and still seeing the same error. Not sure i have a well enough understanding on why this happens. I can see there is another slot which defines:
```
    @pyqtSlot(QWebEnginePage.RenderProcessTerminationStatus, int)
```

Which works.

The-Compiler replied:
Then I'm assuming (like @toofar said) your PyQt, Qt, or the Qt you compiled PyQt against is too old - see `:version` and make sure all Qt version numbers are 5.14 or newer.

rakanalh replied:
This is what i see:
```
Qt: 5.14.2

CPython: 3.8.2
PyQt: 5.14.2

PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.14.0
PyQt5.QtWebKitWidgets: no
```

rakanalh replied:
@The-Compiler Does it make a difference that this signal is defined in `QWebEnginePrivate`?
https://code.qt.io/cgit/qt/qtwebengine.git/tree/src/webenginewidgets/api/qwebenginepage.cpp?h=5.15#n1887

rakanalh commented:
The current version of the code works with a timer to check for recommended lifecycle changes. 

One problem i noticed though, which is that when a tab is discarded, i see QB logging:
```
17:35:18 DEBUG    misc       eventfilter:eventFilter:80 <qutebrowser.browser.webengine.webview.WebEngineView object at 0x7fb562531040>: removed child <PyQt5.QtCore.QObject object at 0x7fb562531ee0>
```
Then once this happens, navigating to the discarded tab would reload but nothing would appear (white background). Navigating away to a different tab and then coming back shows the web content and the whitebackground goes away. I think the removal of the `WebEngineView` has something to do with this... i will have to look for clues on where this happens.

coiby commented:
> The current version of the code works with a timer to check for recommended lifecycle changes.
>
> One problem i noticed though, which is that when a tab is discarded, i see QB logging:
>
> ```
> 17:35:18 DEBUG    misc       eventfilter:eventFilter:80 <qutebrowser.browser.webengine.webview.WebEngineView object at 0x7fb562531040>: removed child <PyQt5.QtCore.QObject object at 0x7fb562531ee0>
> ```
>
> Then once this happens, navigating to the discarded tab would reload but nothing would appear (white background). Navigating away to a different tab and then coming back shows the web content and the whitebackground goes away. I think the removal of the `WebEngineView` has something to do with this... i will have to look for clues on where this happens.

I also encountered this issue when working on #5398. This is caused by a bug in chromium. You can check my PR for the details.

There are more duplicated work between your work and #5398 since you latest work now also use a timer to change a page's state though it's based on the page's recommendedState. [Qt's documents](https://doc.qt.io/qt-5/qwebenginepage.html#recommendedState-prop) don't give much detail on how a page's recommended state is changed,
> Setting the lifecycle state to a lower resource usage state than the recommended state may cause side-effects such as stopping background audio playback or loss of HTML form input. Setting the lifecycle state to a higher resource state is however completely safe.

But I think a separate suspender to automatically discard tabs is necessary, for example,
 - normally we don't want to suspend a pinned tab
 - we never want to discard certain pages, i.e., whitelisting some websites

 I don't think Qt Webengine's internal mechanism of changing recommendedState can handle those cases.

rakanalh commented:
> I also encountered this issue when working on #5398. This is caused by a bug in chromium. You can check my PR for the details.

@The-Compiler  was kind enough to provide this link for the issue:
https://bugreports.qt.io/browse/QTBUG-83641

which is exactly what i was seeing.

> Qt's documents don't give much detail on how a page's recommended state is changed,

What i noticed is that, as soon as a page is not visible, the recommended state changes to `Frozen` and then a second after it is changed again to `Discarded`. The LifecycleState Qt example shows that they've implemented two timers that are started after the recommended state is changed so that the lifecycle can be set on the page after the timers timeout events fire. This is what i also chose to implement because it doesn't make sense to freeze/discard the tabs as soon as they're invisible.

> I don't think Qt Webengine's internal mechanism of changing recommendedState can handle those cases.

Right! Maybe you can implement those ideas on top of this PR as an improvement? there's still a bit more work to do on this PR to make CI green.

coiby commented:
> > Qt's documents don't give much detail on how a page's recommended state is changed,
>
> What i noticed is that, as soon as a page is not visible, the recommended state changes to `Frozen` and then a second after it is changed again to `Discarded`. The LifecycleState Qt example shows that they've implemented two timers that are started after the recommended state is changed so that the lifecycle can be set on the page after the timers timeout events fire. This is what i also chose to implement because it doesn't make sense to freeze/discard the tabs as soon as they're invisible.

Previously I completely ignore the recommended state and discarded tabs based on my own rules. You work and the explanation here makes me realize I should respect the recommended state since it's calculated in a more thoughtful way than I thought.

> > I don't think Qt Webengine's internal mechanism of changing recommendedState can handle those cases.
>
> Right! Maybe you can implement those ideas on top of this PR as an improvement? there's still a bit more work to do on this PR to make CI green.

I'm happy to implement them based on your work! Looking forwarding to you finishing this PR.

toofar commented:
What do you think about not freezing or discarding pinned (or pinned with tabs.pinned.frozen set) tabs? The intent would be re-using the pinned tabs feature to mark tabs as "always active". So that I can leave some tab in the background and expect it to keep updating and notify me when something happens. I'm just re-using `self.navigation_blocked()` in `_check_recommended_lifecycle_state()` currently.

toofar commented:
@rakanalh do you want to rebase or do you want me to?

rakanalh commented:
@toofar Made a best effort merge into the branch. Let me know if you find any issue(s).

coiby commented:
Hi @rakanalh

I noticed a problem about this PR.

1. A tab is discarded
2. Re-focus on the discarded tab, the lab becomes active again
3. Focus on another tab, the previous re-activated tab is discarded immediately

I'm not sure if it's related to the configuration
```
c.tabs.lifecycle_freeze_delay = 60000
c.tabs.lifecycle_discard_delay = 1800000
```

I encounter this issue on a daily basis.

dccsillag commented:
Hey guys,

What's the status on this? Is there anything I can do to help?

The-Compiler commented:
I haven't really had a chance to pick it up so far again. I plan to take another stab at the PR backlog in early 2022 when I have more time, but no promises. I also see @arza-zara picked things up, so they might be able to say more about in which state it is.

dccsillag commented:
Thanks!

Just to be clear, I am willing to help on this, but I don't know how (besides reimplementing everything, but I'd imagine that's undesirable/unnecessary).

toofar commented:
My summary for what it is worth:

1. I'm still happily using it (although I'm pretty sure I still have some local patches)
2. I suspect this won't be the first cab off the rank when more things start getting merged (I think there are some higher value PRs pending yet but go +1 the linked issue if you haven't already)
3. re-implementing (as in re-ordering, squashing etc) is probably going to happen anyway when we get to it (just my opinion!), this PR is getting a bit unwieldy, the diff isn't that bad but there are many commits. I would split it into 1) add more consistently unified session history abstraction for both backends (the webkit one has lagged behind the webengine one IIRC), that's a requirement for this (while we still support both backends) and would make the next change look a lot cleaner if it were done separately (also I have a different change which depends on that) 2) add the lazy loading handling in 3) keep the lifecycle management in a separate PR

toofar commented:
It looks like on Qt6 https://github.com/qutebrowser/qutebrowser/pull/4923 is required again for this to be most effective, or something much like it. As it is now a new renderer process will get created for each and every tab, including the not-yet-loaded ones (wheres on Qt 5.15 the renderer processes only got created when you first focused the tab). The tabs still aren't "loaded", as in there is no HTML rendered in them and no JS run, but they still get renderer processes created, which can be rather weighty depending on your tab hoarding habits.

You can test this by lazy loading a session and looking in :tab-select, ideally all the not-yet-loaded tabs will have a processes ID of `0`. On Qt6 (if you can get that working :smile:) everything has a real processes ID. You can apply parts of #4923 and see how that changes. For example you can add the show()/hide() and background=True bits in sessions.py and see that now only the first and active tabs have real PIDs.

arza-zara commented:
I fixed conflicts and Qt 6 issues, added tests and fixed the tab switching issue. Remaining test failures:
```
FAILED tests/end2end/features/test_sessions_bdd.py::test_saving_with_nohistory - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'category': 'message', 'loglevel': 20, 'message': 'Saved session /tmp/tmp0hf47mgn/session.yml.'}.
FAILED tests/end2end/features/test_sessions_bdd.py::test_saving_with_nohistory_and_onlyactivewindow - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'category': 'message', 'loglevel': 20, 'message': 'Saved session /tmp/tmpiowzipr6/session.yml.'}.
FAILED tests/end2end/features/test_tabs_bdd.py::test_simple_tab_unload_then_go_backward_to_a_new_background_tab - Failed: Session comparison failed: Second list is longer than first list
FAILED tests/end2end/features/test_tabs_bdd.py::test_simple_tab_unload_then_go_forward_to_a_new_background_tab - Failed: Session comparison failed: Second list is longer than first list
FAILED tests/end2end/features/test_tabs_bdd.py::test_tab_unload_and_tab_focus_with_history - Failed: Session comparison failed: Second list is longer than first list
FAILED tests/end2end/features/test_tabs_bdd.py::test_tab_unload_and_tab_move - end2end.fixtures.testprocess.BlacklistedMessageError: 13:39:56.022 DEBUG    webview    tabbedbrowser:_on_title_changed:796 Changing title for idx 1 to 'http://localhost:38083/data/numbers/3.txt'
```
Also, I some time got a tab reloading every time when focusing that tab, and I suspect something wrong with the lifecycle stuff, maybe related to https://github.com/qutebrowser/qutebrowser/pull/4037#issuecomment-772386099.

illfygli commented:
I'm also giving this a look.
It didn't work for me at all, initially, because I use my window manager's tabs, and not qutebrowser's tabs, and there was a visibility check on qute tabs.

When playing around with that, I noticed that this is already taken into account by the recommended state.

Furthermore, there is a signal for a change in recommended state.

So I removed the timers and most of the logic, and just set the state as recommended.

This already takes into account things like HTML form state (so you don't lose data), media playback, and whatnot. So to me this works great thus far.

The only setting needed would then be to use the recommended states or not.

But now it's becoming quite different from this PR, so I don't know if I should open a new one? Or does this seems like a good direction to take this one?

(While writing this in nvim, I noticed how my otherwise always-fanning laptop is now quiet, despite many browser windows, and the CPU monitor itself is the biggest CPU thief.)
